### DISCUSSION TITLE
Support disabling logs

### DISCUSSION THREAD

**Discussion by tanguy-michardiere** [fastapi/fastapi-cli#15] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched in the documentation/README.
- [X] I already searched in Google "How to do X" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs/README and didn't find an answer.

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```Text
`fastapi run`
```

### Description

Hello,

I would like to be able to disable the CLI's logs, mostly the panels when it starts.
The cherry-on-top would be to be able to disable uvicorn's logs too ("Uvicorn running on ...").
Ideally it could be configured with environment variables.

Thank you

### Operating System

macOS

### Operating System Details

14.4.1 on M3 Max

### Python Version

3.12.3

**Comment by quswadress** [comment]
This would be funny if it weren't so sad. FastAPI positions itself as a high-performance framework, but its nice ‘rich’ output cannot be disabled or redirected to a file (or null) in Windows.
```
C:\>fastapi --help > lol.txt
Traceback (most recent call last)
│ D:\PycharmProjects\web_bench\.venv\lib\site-packages\rich\_win32_console.py │
│ :403 in write_text                                                          │
│                                                                             │
│   400 │   │   Args:                                                         │
│   401 │   │   │   text (str): The text to write to the console              │
│   402 │   │   """                                                           │
│ > 403 │   │   self.write(text)                                              │
│   404 │   │   self.flush()                                                  │
│   405 │                                                                     │
│   406 │   def write_styled(self, text: str, style: Style) -> None:          │
│                                                                             │
│ ┌──────────────────────────────── locals ─────────────────────────────────┐ │
│ │ self = <rich._win32_console.LegacyWindowsTerm object at                 │ │
│ │        0x00000237A2993A30>                                              │ │
│ │ text = ' command line app. 😎'                                          │ │
│ └─────────────────────────────────────────────────────────────────────────┘ │
└─────────────────────────────────────────────────────────────────────────────┘
UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f60e' in
position 19: character maps to <undefined>
```
With one line `if False:` in `rich\_win32_console.py` I increased the number of requests per second by 2-3 times:
```
PS D:\PycharmProjects\web_bench> echo "FastApi in different cmd"
PS D:\web_bench> .\bombardier-windows-amd64.exe -c 4 -d 60s http://127.0.0.1:8000/
Bombarding http://127.0.0.1:8000/ for 1m0s using 4 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      4246.71      81.69    4495.60
  Latency        0.94ms    33.40us     3.38ms
  HTTP codes:
    1xx - 0, 2xx - 254798, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   850.15KB/s
PS D:\web_bench> .\bombardier-windows-amd64.exe -c 16 -d 60s http://127.0.0.1:8000/
Bombarding http://127.0.0.1:8000/ for 1m0s using 16 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      8153.03    1055.87    9373.41
  Latency        1.96ms   701.53us    45.47ms
  HTTP codes:
    1xx - 0, 2xx - 489071, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     1.59MB/s
PS D:\web_bench> echo "FastApi with my patch and 2>&1>$null"
PS D:\web_bench> .\bombardier-windows-amd64.exe -c 4 -d 60s http://127.0.0.1:8000/
Bombarding http://127.0.0.1:8000/ for 1m0s using 4 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      4110.98     125.68    7404.66
  Latency        0.97ms    39.81us     4.47ms
  HTTP codes:
    1xx - 0, 2xx - 246598, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   822.78KB/s
PS D:\web_bench> .\bombardier-windows-amd64.exe -c 16 -d 60s http://127.0.0.1:8000/
Bombarding http://127.0.0.1:8000/ for 1m0s using 16 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec     27319.84     609.04   28754.17
  Latency      582.79us    12.20us     4.00ms
  HTTP codes:
    1xx - 0, 2xx - 1639203, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     5.34MB/s
```

**Comment by williamweckl** [comment]
I am having some trouble with this subject too.

I have an implementation that is based on [this gist](https://gist.github.com/nymous/f138c7f06062b7c43c060bf03759c29e), and after changing to use `fastapi run` instead of starting directly the uvicorn I see double logs for requests because I can't disable the ones from fastapi cli.

What I want to do is basically the same that I was doing with unicorn using the lines bellow:
```python
logging.getLogger("uvicorn.access").handlers.clear()
logging.getLogger("uvicorn.access").propagate = False
```

Does anyone know if this is possible?

**Reply by williamweckl** [reply]
Looks like setting up the logger changes insite lifespan fixes the issue, at least the one that I was dealing with.

```python
@asynccontextmanager
async def lifespan(_app: FastAPI):
    setup_logging(log_level=LOG_LEVEL)
    yield

app = FastAPI(lifespan=lifespan)
```

**Reply by lraubuch** [reply]
This is the right way to configure the logging for uvicorn since it has to be done after the app was created and the lifespan is the perfect place for it 👍🏻

**Reply by LeoVasanko** [reply]
Note that in order to get startup messages under control, you have to pass in custom logging config to `uvicorn.run` which is not possible to do by FastAPI cli. If you are looking for better control of your app startup, it is better to use that function directly.

**Comment by rbngz** [comment]
By default, the `uvicorn` and `uvicorn.access` loggers have custom handlers which are used to log messages. The `uvicorn.access` logger does not propagate by default to avoid having duplicate logs from `uvicorn` and `uvicorn.access`.
The `uvicorn.error` logger has no custom handler and propagates to the `uvicorn` logger by default

While it is possible to disable "runtime" logs using 

```python
# Clear custom rich handlers of uvicorn and uvicorn.access
logging.getLogger("uvicorn.access").handlers.clear() 
logging.getLogger("uvicorn").handlers.clear()

# Option A: disable propagation of all uvicorn logs
logging.getLogger("uvicorn").propagate = False

# Option B: propagate uvicorn logs to root logger
logging.getLogger("uvicorn.access").propagate = True # Does not propagate by default
```

It is not possible to disable the start up logs that appear immediately after running `fastapi run` or `fastapi dev`. This is especially annoying with serverless containers that will show the same startup logs everytime the instance has been idle for some time, or a new instance is started up. Would be great to be able to pass a CLI argument that disables such logs.

**Comment by eden881** [comment]
Any update on this? We really need this so we can stop poisoning our log analytics 😃

**Comment by CostcoFanboy** [comment]
An update would be nice. Structured logging is pretty essential for any production environment (e.g. Kibana, Splunk, Datadog).

The easy way would be to pass the flag to uvicorn just to mimic functionality:
`fastapi run main.py --log-config logging.yaml`

But it would be incredibly nice for popular powerful 3rd party logging libraries like loguru and structlog to work seamlessly with FastAPI-CLI. 

So I understand that disabling it entirely feels like it should be an option.

```fastapi run main.py --no-rich-logging```

But perhaps, if possible, making it a callable setup would be even more baller:

```fastapi run main.py --log-setup app.logging:setup_logging```

**Comment by MarleTangible** [comment]
I'm trying to test some code that processes 20 requests per second and the terminal output is just insane with access logs, there should be a way to disable this on Windows, I need to see the errors only not a wall of INFO messages.
